internal/provider/string.go

```diff
@@ -1,219 +1,382 @@
+				Description: "Minimum number of special characters in the result. Default value is `0`.",
+				Type:        types.Int64Type,
+				Optional:    true,
+				Computed:    true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{
+					tfsdk.RequiresReplace(),
+					defaultInt(0),
+				},
+			},
 
-		"override_special": {
-			Description: "Supply your own list of special characters to use for string generation.  This " +
-				"overrides the default character list in the special argument.  The `special` argument must " +
-				"still be set to true for any overwritten characters to be used in generation.",
-			Type:     schema.TypeString,
-			Optional: true,
-			ForceNew: true,
-		},
+			"override_special": {
+				Description: "Supply your own list of special characters to use for string generation.  This " +
+					"overrides the default character list in the special argument.  The `special` argument must " +
+					"still be set to true for any overwritten characters to be used in generation.",
+				Type:     types.StringType,
+				Optional: true,
+				Computed: true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{
+					tfsdk.RequiresReplace(),
+				},
+			},
 
-		"result": {
-			Description: "The generated random string.",
-			Type:        schema.TypeString,
-			Computed:    true,
-			Sensitive:   sensitive,
-		},
+			"result": {
+				Description: "The generated random string.",
+				Type:        types.StringType,
+				Computed:    true,
+				Sensitive:   sensitive,
+			},
 
-		"id": {
-			Description: idDesc,
-			Computed:    true,
-			Type:        schema.TypeString,
+			"id": {
+				Description: idDesc,
+				Computed:    true,
+				Type:        types.StringType,
+			},
 		},
 	}
 }
 
-func createStringFunc(sensitive bool) func(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	return func(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-		const numChars = "0123456789"
-		const lowerChars = "abcdefghijklmnopqrstuvwxyz"
-		const upperChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
-		var (
-			specialChars = "!@#$%&*()-_=+[]{}<>:?"
-			diags        diag.Diagnostics
+func validatorMinInt(min int64) tfsdk.AttributeValidator {
+	return minIntValidator{min}
+}
+
+type minIntValidator struct {
+	val int64
+}
+
+func (m minIntValidator) Description(context.Context) string {
+	return "MinInt validator ensures that attribute is at least val"
+}
+
+func (m minIntValidator) MarkdownDescription(context.Context) string {
+	return "MinInt validator ensures that attribute is at least `val`"
+}
+
+func (m minIntValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, resp *tfsdk.ValidateAttributeResponse) {
+	t := req.AttributeConfig.(types.Int64)
+
+	if t.Value < m.val {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("expected attribute to be at least %d, got %d", m.val, t.Value),
+			fmt.Sprintf("expected attribute to be at least %d, got %d", m.val, t.Value),
 		)
+	}
+}
```

Comment:
Nitpick: Would it make sense to have the Validators and PlanModifiers into their own files?
Like a `validators.go` and a `plan_modifiers.go`?

I might be wrong but they feel a bit "out of their context" in here, given that some are also used in other contexts (not just `random_string` and `random_password`).

Reply:
Personally, I agree with the above. It'll also help us as we consider making a corpus of "common" validators in the near future.

Reply:
Makes sense to me. Have moved `validators` and `plan modifiers` to separate files.